src/Microsoft.VisualStudio.Threading/AsyncSemaphore.cs

```diff
@@ -97,7 +97,14 @@ protected virtual void Dispose(bool disposing)
         {
             if (disposing)
             {
-                this.semaphore.Dispose();
+                // !!DO NOT!! dispose the semaphore, for the following reasons:
+                // 1. SemaphoreSlim.Dispose is not safe for concurrent use while something is waiting. Calling Dispose
+                //    requires tracking of all outstanding requests for enter/release.
+                // 2. SemaphoreSlim only allocates resources that could need to be be disposed if the
+                //    SemaphoreSlim.AvailableWaitHandle property is accessed. The semaphore field here is private and we
+                //    do not access this property.
```

Comment:
@AArnott If it always threw, then that would be more workable. However, the behavior in this scenario is unspecified. It might or might not throw, the type of exception could vary, and the exception may or may not occur in any specific call. Maybe the first invalid call gets an exception, or maybe something down the line does.